### PR TITLE
VideoCommon: Eliminate static state in Renderer

### PR DESCRIPTION
**Pull request by stenzek**
This has been bugging me for a while.

Also adds a cache for the last window size request, this way it avoids posting a message every frame (at least in WX, anyway).

Probably needs some more testing, I did a really brief check with all backends and it seemed okay, but due to the init order changing in some cases there's a slim possibility of issues.

### CODE REVIEW

**Comment by ligfx** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/VideoBackends/D3D/main.cpp [hunk 1]
```diff
@@ -144,11 +144,15 @@ bool VideoBackend::Initialize(void* window_handle)
 
 void VideoBackend::Video_Prepare()
 {
+  if (FAILED(D3D::Create(reinterpret_cast<HWND>(m_window_handle))))
+    PanicAlert("Failed to create D3D device.");
```
Comment (lioncash) [inline Source/Core/VideoBackends/D3D/main.cpp]: `"Common/MsgHandler.h"` needs to be included.

File: Source/Core/VideoBackends/Software/SWRenderer.h [hunk 2]
```diff
@@ -13,7 +13,8 @@
 class SWRenderer : public Renderer
 {
 public:
-  ~SWRenderer() override;
+  SWRenderer();
+  ~SWRenderer();
```
Comment (lioncash) [inline Source/Core/VideoBackends/Software/SWRenderer.h]: why was the `override` specifier removed?

File: Source/Core/VideoCommon/RenderBase.h [hunk 3]
```diff
@@ -79,48 +79,52 @@ class Renderer
   virtual void RestoreState() {}
   virtual void ResetAPIState() {}
   virtual void RestoreAPIState() {}
+  // Some of the methods called by here assume g_renderer is initialized, therefore
```
Comment (lioncash) [inline Source/Core/VideoCommon/RenderBase.h]: This should probably have a TODO or FIXME on it. This is very indicative of an initialization order issue or dependencies being where they shouldn't be.

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 4]
```diff
@@ -108,20 +76,30 @@ static float AspectToWidescreen(float aspect)
...
+
   OSDChoice = 0;
   OSDTime = 0;
 }
 
 Renderer::~Renderer()
 {
   // invalidate previous efb format
-  prev_efb_format = PEControl::INVALID_FMT;
+  m_prev_efb_format = PEControl::INVALID_FMT;
```
Comment (lioncash) [inline Source/Core/VideoCommon/RenderBase.cpp]: Given these are now non-static, is this and the below line still necessary? If the below code requires it, this should probably be moved into `ShutdownFrameDumping()`

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 5]
```diff
@@ -108,20 +76,30 @@ static float AspectToWidescreen(float aspect)
...
 
-Renderer::Renderer()
+Renderer::Renderer(int backbuffer_width, int backbuffer_height)
 {
+  FramebufferManagerBase::SetLastXfbWidth(MAX_XFB_WIDTH);
+  FramebufferManagerBase::SetLastXfbHeight(MAX_XFB_HEIGHT);
+
   UpdateActiveConfig();
 
+  m_backbuffer_width = backbuffer_width;
```
Comment (lioncash) [inline Source/Core/VideoCommon/RenderBase.cpp]: width and height can likely be set in the constructor initializer list, assuming none of the preceding function calls do weird shenanigans with the class variables (if they do, a comment here to explain that might be helpful).

File: Source/Core/VideoBackends/Software/SWRenderer.h (hunk 2 above)
Comment (stenzek) [inline Source/Core/VideoBackends/Software/SWRenderer.h]: This was what I pinged you on IRC about, which I understand the reasoning for now :). I've also added the specifier to the destructors for the other derived Renderer classes.

File: Source/Core/VideoCommon/RenderBase.cpp (hunk 4 above)
Comment (stenzek) [inline Source/Core/VideoCommon/RenderBase.cpp]: Nice catch. Done.

File: Source/Core/VideoCommon/RenderBase.h (hunk 3 above)
Comment (stenzek) [inline Source/Core/VideoCommon/RenderBase.h]: It ended up just being a single method call, which I've relocated elsewhere now, so it's not necessary any more.

**Comment by degasus** [comment]
LGTM. Indeed, there is quite a chance for regressions, but the change is very worth it.

**Comment by grint87** [comment]
Been running development builds for some time, this is the first one which completely broke on all backends :(

**Comment by degasus** [comment]
In which way is it broken? Writing a fix is likely easy just by the description of the bug.

**Comment by grint87** [comment]
I'll get back to you after some testing.

**Comment by grint87** [comment]
Okay, tested Luigi's Mansion with OpenGL backend on Windows 10 with AMD R9 380 latest drivers. Seems as through Native resolution is fine, and so is 2x Native for 720p however 3x Native for 1080p seems to immediately crash with "Dolphin.exe has stopped working". It's the same story with other GameCube titles - didn't get around to Wii emulation.

**Pull request merged**